src/App/Fossa/BinaryDeps/Jar.hs

```diff
@@ -0,0 +1,106 @@
+--   2. Search inside for a file named `pom.xml`; if there are multiple pick the one with the shortest path.
+--      If a representative pom.xml was found, parse it and return metadata derived from it.
+--   3. Attempt to open `META-INF/MANIFEST.MF`, parse it, and return metadata derived from it.
+resolveJar :: (Has (Lift IO) sig m, Has Diagnostics sig m, Has Logger sig m, Has ReadFS sig m) => Path Abs Dir -> Path Abs File -> m (Maybe SourceUserDefDep)
+resolveJar _ file | not $ fileHasSuffix file [".jar", ".aar"] = pure Nothing
+resolveJar root file = do
+  let fileDescription = toText file
+  logDebug $ "Inferring metadata from " <> pretty fileDescription
+  result <- recover . context ("Infer metadata from " <> fileDescription) . runFinally $ withArchive extractZip file $ \dir -> tacticPom dir <||> tacticMetaInf dir
+  pure $ fmap (toUserDefDep root file) result
+
+tacticMetaInf :: (Has (Lift IO) sig m, Has Diagnostics sig m, Has Logger sig m, Has ReadFS sig m) => Path Abs Dir -> m JarMetadata
+tacticMetaInf archive = context ("Parse " <> toText metaInfPath) $ do
+  content <- readContentsText metaInfPath
+  logDebug $ "Parsing META-INF manifest: " <> pretty (renderRelative archive metaInfPath)
+  metaInfManifestToMeta $ parseMetaInfManifest content
+  where
+    metaInfPath = archive </> $(mkRelDir "META-INF") </> $(mkRelFile "MANIFEST.MF")
+
+parseMetaInfManifest :: Text -> Map Text Text
+parseMetaInfManifest t = Map.fromList . map strip' . filter' $ map (Text.breakOn ":") (Text.lines t)
+  where
+    null' (a, b) = any Text.null [a, b]
+    strip' (a, b) = (Text.strip a, Text.strip $ Text.tail b)
+    filter' = filter (not . null')
+
+metaInfManifestToMeta :: Has Diagnostics sig m => Map Text Text -> m JarMetadata
+metaInfManifestToMeta manifest =
+  JarMetadata
+    <$> fromMaybeText "Missing bundle name" (Map.lookup "Bundle-SymbolicName" manifest <|> Map.lookup "Implementation-Title" manifest)
+    <*> fromMaybeText "Missing implementation version" (Map.lookup "Implementation-Version" manifest)
+    <*> pure "" -- Don't attempt to use Bundle-License; it's a URL and we don't parse it on the backend
+
+tacticPom :: (Has (Lift IO) sig m, Has Diagnostics sig m, Has Logger sig m, Has ReadFS sig m) => Path Abs Dir -> m JarMetadata
+tacticPom archive = context ("Parse representative pom.xml in " <> toText archive) $ do
+  poms <- context "Find pom.xml files" $ walk' (collectFilesNamed "pom.xml") (archive </> $(mkRelDir "META-INF"))
+  if length poms > 1
+    then logDebug $ "Found multiple pom.xml files: " <> pretty (Text.intercalate "; " $ map (renderRelative archive) poms)
+    else pure ()
+  pom <- fromMaybeText "No pom.xml files found" $ choosePom poms
+  logDebug $ "Chose representative pom.xml: " <> pretty (renderRelative archive pom)
+  parsePom pom
+
+choosePom :: [Path Abs File] -> Maybe (Path Abs File)
+choosePom [] = Nothing
+choosePom [pom] = Just pom
+choosePom poms = Just . head $ sortOn (length . toString) poms
```

Comment:
[nit] avoid partial function: `head`. There is a ticket in backlog to remove head, tail, etc. 

https://wiki.haskell.org/Partial_functions

Reply:
I think this whole function could be written as `choosePom = listToMaybe . sortOn (length . toString)`.

Though this usage is obviously safe upon reading, it would be nice if it were safe regardless of context.  That's the reasoning behind not using any partial functions.

Reply:
That makes total sense. Thanks, updated!